### PR TITLE
ENH: Added serialization for IncrementalEmpiricalCovariance

### PR DESCRIPTION
**Pull request by olegkkruglov**
### Description 
* Added `__getstate__` method to `IncrementalEmpiricalCovariance`
* Finalization is called with serialization during `__getstate__` method
* `_need_to_finalize` flag is added to avoid unnecessary call of finalization backend.
* Added test for checking if it works well
* Updated deselected_tests.yaml

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes, if necessary.
- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.
- [x] I have resolved any merge conflicts that might occur with the base branch.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_
- [x] I have added a respective label(s) to PR if I have a permission for that.

### CODE REVIEW

File: onedal/covariance/tests/test_incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/tests/test_incremental_covariance.py:30]: it is correct here. check again please.

**Review by icfaust — COMMENTED** [review]
Good work. I'd like to see getstate and setstate tidied up a bit, and overall some documentation since this is not a straightforward problem. Otherwise good to go.

File: onedal/datatypes/data_conversion.cpp
Comment (olegkkruglov) [reply onedal/datatypes/data_conversion.cpp:84]: Here default layout is changed for the cases where converted numpy array is both C-contagious and F-contagious. It can be happen if the array is of shape (1, n) or (n, 1) (it can be seen in `__setstate__` method here). So it means that from_table/to_table cycle changes layout on oneDAL side. and since currently only C-style onedal tables can be converted to Python (I plan to fix it soon), it prevents multiple pickling. Please see https://github.com/intel/scikit-learn-intelex/pull/1924.

File: onedal/covariance/incremental_covariance.py
Comment (icfaust) [reply onedal/covariance/incremental_covariance.py:79]: It looks like the partial result object only contains these three attributes: https://github.com/intel/scikit-learn-intelex/blob/main/onedal/covariance/covariance.cpp#L138
```suggestion
        p_res = data["_partial_result"]
        data["_partial_result"] = {key:from_table(value) for key, value in p_res.__dict__.items()}
```

File: onedal/covariance/incremental_covariance.py
Comment (icfaust) [reply onedal/covariance/incremental_covariance.py:99]: I worry about checking against .size. Is it always initialized to 0? Or is hasattr a better solution? Not sure.

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (ethanglaser) [inline sklearnex/covariance/tests/test_incremental_covariance.py:175]: I see with our existing pickle tests we are only pickling after fitting, is there a reason to validate pickling on just the object before fitting? is this something that should be extended to other algos?

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (ethanglaser) [inline sklearnex/covariance/tests/test_incremental_covariance.py:193]: is this necessary if we already do it in lines 189-190?

File: onedal/covariance/tests/test_incremental_covariance.py
Comment (ethanglaser) [inline onedal/covariance/tests/test_incremental_covariance.py:1]: Is it necessary to validate this in both onedal4py and sklearnex?

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:79]: since it is not class with dynamic attributes (see https://pybind11.readthedocs.io/en/stable/classes.html) it has no `__dict__` attribute. so the second option is not valid.

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:66]: I'll add some comments but I guess documentation will be updated in separate PR since currently we don't even have any mention of incremental algos in our docs.

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (olegkkruglov) [reply sklearnex/covariance/tests/test_incremental_covariance.py:175]: Just in case and to check we don't get fail after such call. for example it's theoretically possible that there might be errors in __getstate__ and __setstate__ because not all attributes are initialized

File: onedal/covariance/tests/test_incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/tests/test_incremental_covariance.py:1]: onedal4py IncrementalEmpiricalCovariance and sklearnex IncrementalEmpiricalCovariance are different classes with different attributes and it's not completely fair to assume that they are either both serializable or both not. i think it doesn't make things worse to recheck again if everything works well. it might make sense to remove onedal4py part because onedal4py estimator is an attribute of sklearnex estimator. but as for me it is better to have independent and as full as possible test coverage on both sides

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (olegkkruglov) [reply sklearnex/covariance/tests/test_incremental_covariance.py:175]: initially this pickling didn't work, so it allowed me to fix a bug here. might make sense to check if other estimators works in this case, especially such that have C++ attributes (not sure if we have examples other than incremental algos).

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py]: could you clarify please what is wrong with size checking? if the tables from getstate are empty, they are converted to empty list (not sure it is good practice but anyway) and here I try to check if getstate returned empty list to avoid providing it into to_table.

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (olegkkruglov) [reply sklearnex/covariance/tests/test_incremental_covariance.py:193]: it is checked here that finalized estimator is serialized without issues.

Comment (olegkkruglov) [reply sklearnex/covariance/tests/test_incremental_covariance.py:203]: oh, my bad, for sklearnex estimator the finalization occurs after result check. fixed.

**Review by icfaust — COMMENTED** [review]
Since the `__dict__` thing I looked into the pybind11 documentation a little bit more.  I found this: https://pybind11.readthedocs.io/en/stable/advanced/classes.html#pickling-support  Should we do it on the cpp side for the _partial_result object only?  It would be closer to the source in this case rather than doing it on the python side.  You should be able to use `convert_to_pyobject` and `convert_to_table` from the data_conversion header file.

File: onedal/covariance/incremental_covariance.py
Comment (icfaust) [inline onedal/covariance/incremental_covariance.py:91]: Just so I understand, what happens when the _policy attribute doesn't exist?  What does that default to in the computation, or will it get auto-generated with a _get_policy call somewhere later on (after unpickling)?

Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:91]: yes, it will be autogenerated using provided data. because we finalized computation before serialization, the case when we have to call finalize_fit without data is not possible.

**Comment by olegkkruglov** [comment]
> Since the `__dict__` thing I looked into the pybind11 documentation a little bit more. I found this: https://pybind11.readthedocs.io/en/stable/advanced/classes.html#pickling-support Should we do it on the cpp side for the _partial_result object only? It would be closer to the source in this case rather than doing it on the python side. You should be able to use `convert_to_pyobject` and `convert_to_table` from the data_conversion header file.
thanks for pointing out! to be honest I see no serious advantages of this approach over already implemented one. `partial_result` object is used only here by design so no need to implement its serialization separately. If you see something I'm missing please let me know.

**Comment by icfaust** [comment]
> > Since the `__dict__` thing I looked into the pybind11 documentation a little bit more. I found this: https://pybind11.readthedocs.io/en/stable/advanced/classes.html#pickling-support Should we do it on the cpp side for the _partial_result object only? It would be closer to the source in this case rather than doing it on the python side. You should be able to use `convert_to_pyobject` and `convert_to_table` from the data_conversion header file.
> 
> thanks for pointing out! to be honest I see no serious advantages of this approach over already implemented one. `partial_result` object is used only here by design so no need to implement its serialization separately. If you see something I'm missing please let me know.

@samir-nasibli @ethanglaser I'd like one/both of you all to weigh in on this.  I think actions of an object should lie with the object itself if possible.  From a maintenance standpoint, any changes to this pybind11 object will likely be propagated elsewhere to a hard-coded list.  From a usability standpoint, while unexpected, pickling the `_partial_fit` object itself doesn't seem possible?

If you guys like it as is, I will accept it.

File: onedal/covariance/incremental_covariance.py
Comment (samir-nasibli) [inline onedal/covariance/incremental_covariance.py:164]: This comment is out of the scope of the current PR.
I think make sense to do refactoring in follow up. It seems to me that better to have some kind of common BaseIncremenatalEstimator class with such attribute documented well and base abstract methods stored there

File: onedal/covariance/incremental_covariance.py
Comment (samir-nasibli) [inline onedal/covariance/incremental_covariance.py:170]: Another attr for BaseIncremental estimators. Handling it should be on there as well.
https://github.com/intel/scikit-learn-intelex/pull/1925/files#r1677806644

**Comment by icfaust** [comment]
> > > Since the `__dict__` thing I looked into the pybind11 documentation a little bit more. I found this: https://pybind11.readthedocs.io/en/stable/advanced/classes.html#pickling-support Should we do it on the cpp side for the _partial_result object only? It would be closer to the source in this case rather than doing it on the python side. You should be able to use `convert_to_pyobject` and `convert_to_table` from the data_conversion header file.
> > 
> > 
> > thanks for pointing out! to be honest I see no serious advantages of this approach over already implemented one. `partial_result` object is used only here by design so no need to implement its serialization separately. If you see something I'm missing please let me know.
> 
> @samir-nasibli @ethanglaser I'd like one/both of you all to weigh in on this. I think actions of an object should lie with the object itself if possible. From a maintenance standpoint, any changes to this pybind11 object will likely be propagated elsewhere to a hard-coded list. From a usability standpoint, while unexpected, pickling the `_partial_fit` object itself doesn't seem possible?
> 
> If you guys like it as is, I will accept it.

@olegkkruglov I am waiting for a response from them before the next review cycle.

**Comment by icfaust** [comment]
@olegkkruglov any update on this PR? It will need to be rebased for 2025.0 (and what about moving the serialization to pybind side?)

**Comment by icfaust** [comment]
I'm going to wait on private CI to verify the re-enabled test, when it goes through we can merge this.

**Review by icfaust — APPROVED** [review]

File: onedal/covariance/tests/test_incremental_covariance.py
Comment (icfaust) [inline onedal/covariance/tests/test_incremental_covariance.py:148]:
```suggestion
    # Check that estimator can be serialized after partial_fit call.
```

**Review by icfaust — COMMENTED** [review]
please check and make changes as per what is requested in #2180 before merging.

File: onedal/covariance/covariance.cpp
Comment (samir-nasibli) [inline onedal/covariance/covariance.cpp:21]: Why is this needed?

File: onedal/covariance/incremental_covariance.py
Comment (samir-nasibli) [reply onedal/covariance/incremental_covariance.py:65]: We now have this section. I hope this activity already covered by ticket. I am ok to do it in separate PR

**Review by samir-nasibli — APPROVED** [review]

File: onedal/covariance/covariance.cpp
Comment (olegkkruglov) [reply onedal/covariance/covariance.cpp:21]: https://github.com/numpy/numpy/issues/9309#issuecomment-311320497
without this define it fails

**Pull request merged**
